File: src/csharp/JsonValue.cs
using System;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace common;

/// <summary>
/// Provides extension methods for working with <see cref="JsonValue"/> instances in a functional style.
/// </summary>
public static class JsonValueModule
{
    /// <summary>
    /// Safely converts a <see cref="JsonValue"/> to a string.
    /// </summary>
    /// <param name="jsonValue">The <see cref="JsonValue"/> to convert.</param>
    /// <returns>Success with the <see cref="string"/> value if <see cref="JsonValue"/> contains a string, otherwise error details.</returns>
    public static Result<string> AsString(this JsonValue? jsonValue) =>
        jsonValue?.GetValueKind() switch
        {
            JsonValueKind.String => jsonValue.GetStringValue() switch
            {
                null => Result.Success("JSON value has a null string."),
                var stringValue => stringValue
            },
            _ => Error.From("JSON value is not a string.")
        };
        
    private static string? GetStringValue(this JsonValue? jsonValue) =>
        jsonValue?.GetValue<object>().ToString();

    /// <summary>
    /// Safely converts a <see cref="JsonValue"/> to an integer.
    /// </summary>
    /// <param name="jsonValue">The <see cref="JsonValue"/> to convert.</param>
    /// <returns>Success with the <see cref="int"/> value if <see cref="JsonValue"/> contains a valid integer, otherwise error details.</returns>
    public static Result<int> AsInt(this JsonValue? jsonValue)
    {
        var errorMessage = "JSON value is not an integer.";

        return jsonValue?.GetValueKind() switch
        {
            JsonValueKind.Number => int.TryParse(jsonValue.GetStringValue(), out var result)
                                    ? Result.Success(result)
                                    : Error.From(errorMessage),
            _ => Error.From(errorMessage)
        };
    }

    /// <summary>
    /// Safely converts a <see cref="JsonValue"/> to a boolean.
    /// </summary>
    /// <param name="jsonValue">The <see cref="JsonValue"/> to convert.</param>
    /// <returns>Success with the <see cref="bool"/> value if <see cref="JsonValue"/> contains a boolean, otherwise error details.</returns>
    public static Result<bool> AsBool(this JsonValue? jsonValue) =>
        jsonValue?.GetValueKind() switch
        {
            JsonValueKind.True => Result.Success(true),
            JsonValueKind.False => false,
            _ => Error.From("JSON value is not a boolean.")
        };

    /// <summary>
    /// Safely converts a <see cref="JsonValue"/> to a GUID.
    /// </summary>
    /// <param name="jsonValue">The <see cref="JsonValue"/> to convert.</param>
    /// <returns>Success with the <see cref="Guid"/> value if <see cref="JsonValue"/> contains a valid GUID string, otherwise error details.</returns>
    public static Result<Guid> AsGuid(this JsonValue? jsonValue)
    {
        var errorMessage = "JSON value is not a GUID.";

        return jsonValue.AsString()
                        .Bind(stringValue => Guid.TryParse(jsonValue.GetStringValue(), out var result)
                                                ? Result.Success(result)
                                                : Error.From(errorMessage))
                        .MapError(_ => errorMessage);
    }

    /// <summary>
    /// Safely converts a <see cref="JsonValue"/> to an absolute URI.
    /// </summary>
    /// <param name="jsonValue">The <see cref="JsonValue"/> to convert.</param>
    /// <returns>Success with the <see cref="Uri"/> value if <see cref="JsonValue"/> contains a valid absolute URI string, otherwise error details.</returns>
    public static Result<Uri> AsAbsoluteUri(this JsonValue? jsonValue)
    {
        var errorMessage = "JSON value is not an absolute URI.";

        return jsonValue.AsString()
                        .Bind(stringValue => Uri.TryCreate(jsonValue.GetStringValue(), UriKind.Absolute, out var result)
                                                ? Result.Success(result)
                                                : Error.From(errorMessage))
                        .MapError(_ => errorMessage);
    }
}

File: src/fsharp/JsonValue.fs
﻿[<RequireQualifiedAccess>]
module common.JsonValue

open System
open System.Text.Json.Nodes
open System.Text.Json
open FSharpPlus

let private getString (jsonValue: JsonValue) = jsonValue.GetValue<obj>() |> string

let asString (jsonValue: JsonValue) =
    match jsonValue.GetValueKind() with
    | JsonValueKind.String -> getString jsonValue |> JsonResult.succeed
    | _ -> JsonResult.failWithMessage "JSON value is not a string"

let asInt (jsonValue: JsonValue) =
    match jsonValue.GetValueKind() with
    | JsonValueKind.Number ->
        match getString jsonValue |> Int32.TryParse with
        | true, x -> JsonResult.succeed x
        | _ -> JsonResult.failWithMessage "JSON value is not an integer"
    | _ -> JsonResult.failWithMessage "JSON value is not a number"

let asAbsoluteUri jsonValue =
    let errorMessage = "JSON value is not an absolute URI."

    asString jsonValue
    |> bind (fun stringValue ->
        match Uri.TryCreate(stringValue, UriKind.Absolute) with
        | true, uri when
            (match uri with
             | Null -> false
             | NonNull nonNullUri -> nonNullUri.HostNameType <> UriHostNameType.Unknown)
            ->
            JsonResult.succeed uri
        | _ -> JsonResult.failWithMessage errorMessage)
    |> JsonResult.setErrorMessage errorMessage

let asGuid jsonValue =
    let errorMessage = "JSON value is not a GUID."

    asString jsonValue
    |> bind (fun stringValue ->
        match Guid.TryParse(stringValue) with
        | true, guid -> JsonResult.succeed guid
        | _ -> JsonResult.failWithMessage errorMessage)

let asBool (jsonValue: JsonValue) =
    match jsonValue.GetValueKind() with
    | JsonValueKind.True -> JsonResult.succeed true
    | JsonValueKind.False -> JsonResult.succeed false
    | _ -> JsonResult.failWithMessage "JSON value is not a boolean."

let asDateTimeOffset (jsonValue: JsonValue) =
    let errorMessage = "JSON value is not a date time offset."

    match asString jsonValue with
    | JsonResult.Success stringValue ->
        match DateTimeOffset.TryParse(stringValue) with
        | true, dateTimeOffset -> JsonResult.succeed dateTimeOffset
        | _ -> JsonResult.failWithMessage errorMessage
    | JsonResult.Failure _ -> JsonResult.failWithMessage errorMessage
